### PR TITLE
Fix timing point truncation causing missnaps on compatibility-exported lazer beatmaps

### LINKED ISSUE #30606: Kiai time offset can sometimes appear to be off by 1ms on compatibility-exported lazer beatmaps
### Discussed in https://github.com/ppy/osu/discussions/30595

<div type='discussions-op-text'>

<sup>Originally posted by **felipemarins** November 13, 2024</sup>
As the title says, sometimes, when you export a lazer beatmap to .osz, the kiai time may be off by 1ms. I suppose this is due to timing point's precision that is not supported by stable.

Here are some beatmap files where it occurs (one is the .osz, and the other is a .olz for comparison):
[beatmaps-for-testing.zip](https://github.com/user-attachments/files/17727361/beatmaps-for-testing.zip)

Edit: It seems that it happens to hit objects too (but it doesn't trigger AiMod in this case).</div>

---

The problem seems to be that we are truncating the offset of the uninherited timing point without taking into account that it changes the snapping for all objects that come after it. This isn't fixed by rounding instead of truncating.

We'll need to offset all the objects in the timing section by the offset due to truncating the uninherited timing point, and then truncate the objects themselves. For example if the uninherited point is at 1000.4 ms and an object is at 2000.2 ms, we need to first move the object back 0.4 ms to 1999.8 ms and then truncate it to 1999 ms.

_Originally posted by @OliBomby in https://github.com/ppy/osu/discussions/30595#discussioncomment-11238775_

### PR DESCRIPTION
**Pull request by OliBomby**
Fixes https://github.com/ppy/osu/issues/30606

### CODE REVIEW

File: osu.Game.Tests/Beatmaps/IO/LegacyBeatmapExporterTest.cs [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+using osu.Game.Beatmaps;
+using osu.Game.Database;
+using osu.Game.Tests.Resources;
+using osu.Game.Tests.Visual;
+using MemoryStream = System.IO.MemoryStream;
+
+namespace osu.Game.Tests.Beatmaps.IO
+{
+    [HeadlessTest]
+    public partial class LegacyBeatmapExporterTest : OsuTestScene
```
Comment (peppy) [inline osu.Game.Tests/Beatmaps/IO/LegacyBeatmapExporterTest.cs:20]: Just wondering if this test would be better suited to the `LegacyBeatmapEncoderTest` class?

Comment (bdach) [reply osu.Game.Tests/Beatmaps/IO/LegacyBeatmapExporterTest.cs:20]: I'd actually prefer them to be here. Encoder shouldn't be caring about backcompat logic, encoder should be encoding, i.e. dumb formatting logic. This is where I'd also put legacy export stability tests that you asked for in relation to BSS and which we don't currently have.

**Comment by bdach** [comment]
Aside from slight alterations to the assertions in the test (8df7a6f2f31226dab8e1a1d300968b527e24c00b) I see no fault with this.

@peppy anything here you want to intervene on or am I fine to proceed?

**Review by bdach — APPROVED** [review]

**Comment by peppy** [comment]
> Aside from slight alterations to the assertions in the test (8df7a6f2f31226dab8e1a1d300968b527e24c00b) I see no fault with this.
> 
> 
> 
> @peppy anything here you want to intervene on or am I fine to proceed?

at your discretion

**Pull request merged**